keps/sig-node/20191002-seccomp-custom-profiles.md

```diff
@@ -0,0 +1,351 @@
+
+ ```golang
+ // ContainerStatus represents the status of a container.
+type ContainerStatus struct {
+	...
+	// Status of the container.
+	SecurityContext SecurityContextStatus
+	...
+}
+
+ // SecurityContextStatus represents the security context status of a container.
+type SecurityContextStatus struct {
+	// SeccompProfileHash of the SeccompProfile applied.
+	SeccompProfileHash string
+}
+ ```
+
+A similar approach would also be beneficial to AppArmor profiles. 
+
+
+#### Rollout of profile changes
+
+Seccomp profiles are applied at container creation time, therefore updating an existing user-defined profile will not cause any changes to the running containers that are using it. 
+They will need to be restarted in order for the changes to take effect, which users will have to do manually. 
+
+However, the recommended approach for making profile changes is to create a new `ConfigMap` with the desired profile and then trigger a rolling-update, by changing the configmap the pods are pointing to. The old `ConfigMap` can be deleted once there are no running pods referring to it.
+
+
+#### Starting containers with non-existent profile
+
+When the Kubelet is creating a container using this seccomp profile type, it must validate that the profile is valid. For a profile reference of `configmap/webapi-seccomp/profile-block.json`, the Kubelet will check for the existence of a `ConfigMap` named `web-seccomp` in the same namespace as the pod/container referencing it. It will also check whether that `ConfigMap` contains a file named `profile-block.json` and that it is a valid json. If any of those evaluations are false, the container will not be created and an error should be logged instead. 
+
+Once the container is running, the `ConfigMap` holding the profile will not be re-evaluated. If that `ConfigMap` is deleted at this point, nothing will happen until the container is restarted, in which case the same rules as above will apply, leading to the container failing to start.
+
+The Kubelet will only validate whether the contents of the profile file is a valid `json` file. The container runtime will be responsible for validating it further. The Kubelet will not validate or use any other files apart from the one being referenced.
+
+
+### User Stories
+
+#### Story 1
+As a user, I want to be able to create and maintain custom seccomp profiles without having to physically copy files across all my existing (and new) nodes.
+
+
+#### Story 2
+As a user, I want to be able to create and maintain custom seccomp profiles by using kubectl, in the same way I do other Kubernetes objects. 
+
+
+#### Story 3
+As a user, I want to be able to determine whether custom seccomp profiles have been changed since applied to running containers. 
+
+
+
+## Design Details
+
+### No persistence to disk
+
+The profiles will not persisted to disk. The Kubelet will fetch the contents from `ConfigMap` and then returned to be later pass on to the OCI, as per code below:
```

Comment:
```suggestion
The profiles will not persisted to disk. The Kubelet will fetch the contents from `ConfigMap` and then returned to be later pass on to the CRI, as per code below:
```

Do we really want to pass the whole JSON blob via the CRI? Seccomp profiles can be large (multiple kilobytes) from my point of view.

I'm also missing the scope what the container runtime should do with the profile information. If the whole profile gets passed via the CRI every time a workload starts, then we do not have to persist the profile state on disk, right?

Reply:
@saschagrunert Seccomp profiles can indeed grow large, depending on their targeted workload. The motivation to pass them in memory were primarily to have a simple implementation - like the one already in place for dockershim - whilst still overcoming the need to manage files or keep them in sync across nodes.

> I'm also missing the scope what the container runtime should do with the profile information. If the whole profile gets passed via the CRI every time a workload starts, then we do not have to persist the profile state on disk, right?

The initial idea was to send the profile on CreateContainerRequest and RunPodSandbox, which is when/home the filePath is currently being sent for "localhost/" seccomps.

Reply:
Okay I see your arguments and will answer on the main thread. Let's close this discussion for now.